src/ray/object_manager/plasma/connection.h

```diff
@@ -37,12 +38,48 @@ class Client : public ray::ClientConnection, public ClientInterface {
 
   const std::unordered_set<ray::ObjectID> &GetObjectIDs() override { return object_ids; }
 
-  virtual void MarkObjectAsUsed(const ray::ObjectID &object_id) override {
-    object_ids.insert(object_id);
+  // Holds the object ID. If the object ID has a fallback-allocated fd, adds the ref count
+  // to that fd. Note: used_fds_ is not updated. rather, it's updated in SendFd().
+  //
+  // Idempotency: only increments ref count if the object ID was not held.
+  // TODO(ryw): if this method is called with same object_id but different fd, the latter
+  // fd is ignored. Is this possible? Is it OK?
+  virtual void MarkObjectAsUsed(
+      const ray::ObjectID &object_id,
+      std::optional<MEMFD_TYPE> fallback_allocated_fd) override {
+    const auto [_, inserted] = object_ids.insert(object_id);
+    if (fallback_allocated_fd.has_value() && inserted) {
+      MEMFD_TYPE fd = fallback_allocated_fd.value();
+      object_ids_to_fds_[object_id] = fd;
+      fds_ref_count_[fd] += 1;
+    }
   }
 
-  virtual void MarkObjectAsUnused(const ray::ObjectID &object_id) override {
-    object_ids.erase(object_id);
+  // Removes object ID's ref to the fd, if any. If the fd's refcnt goes to 0, also remove
+  // the fd from used_fds_ and return true, directing the client to unmap it.
+  //
+  // Returns: bool, client should unmap.
+  // Idempotency: only decrements ref count if the object ID was held.
+  virtual bool MarkObjectAsUnused(const ray::ObjectID &object_id) override {
+    size_t erased = object_ids.erase(object_id);
+    if (erased == 0) {
+      return false;
+    }
+    auto iter = object_ids_to_fds_.find(object_id);
+    if (iter == object_ids_to_fds_.end()) {
+      return false;
+    }
+    MEMFD_TYPE fd = iter->second;
+    object_ids_to_fds_.erase(iter);
+
+    size_t &ref_count = fds_ref_count_[iter->second];
```

Comment:
if fds_ref_count_[iter->second]; didn't exist before (this will create a new entry if it didn't exist), it can have a leak (because ref_count == 0 and you decrement - 1).

Reply:
Maybe we can have a bit stronger check since the state between `fds_ref_count_` and `object_ids_to_fds_` should be in sync ? (I beleive it is still idepmotent)

```
auto it = fds_ref_count_.find(fd)
// If fd existed before from object_ids_to_fds_ the ref count should have been > 0
RAY_CHECK(it != fds_ref_count_.end());
size_t &ref_cnt = fds_ref_count_[fd];
RAY_CHECK(ref_cnt > 0);
fds_ref_count_[fd] -= 1
if (ref_count == 0) {
...
}
```

Reply:
done.